org/mozilla/jss/netscape/security/x509/CertificateChain.java

```diff
@@ -150,6 +153,85 @@ private void readObject(java.io.ObjectInputStream in)
         decode(in);
     }
 
+    public void addCertificate(X509Certificate cert) {
+        certs.add(cert);
+    }
+
+    public void addCertificateChain(CertificateChain certChain) {
+        certs.addAll(certChain.certs);
+    }
+
+    public void addPKCS7(PKCS7 pkcs7) {
+        certs.addAll(Arrays.asList(pkcs7.getCertificates()));
```

Comment:
I hate to be a spoilsport, but I don't think this is strictly true all the time. For instance, CA bundles are usually a concat'd list of PKCS#7 format PEM certificates (with headers/footers & the works), but they're not strictly a _chain_ (with a connected path from root -> end-entity cert). For what you're doing in https://github.com/dogtagpki/pki/pull/3459, this will suffice, but in general, it isn't strictly true that a PKCS#7 will always contain a strict chain of certs.

Where I'm going with this is, you might want to decide: do you want to validate the contents of the PKCS#7 and ensure it is a single path from root -> leaf, or do you want to ignore the problem and hand-wave it away (does it break sorting?) or what. :-)

My 2c.

Reply:
OK, I've updated the patch to sort (and thus validate) the cert chain so it will only accept a single chain and reject a CA bundle. This patch will still be used by the `pki pkcs7` tools in dogtagpki/pki#3459, but for OCSP installation I think we can get the CA cert from subsystem cert chain instead of the CA bundle, so we don't have to deal with multiple chains. We can also try to remove the CA cert from OCSP's CRL issuing point record as we discussed on IRC, but I think that can be solved separately later.

Reply:
ty! :)